spangres/toddl.go

```diff
@@ -0,0 +1,239 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spangres
+
+import (
+	"fmt"
+	"strconv"
+	"unicode"
+
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+// schemaToDDL performs schema conversion from the source DB schema to
+// Spanner. It uses the source schema in conv.SrcSchema, and writes
+// the Spanner schema to conv.SpSchema.
+func SchemaToDDL(conv *internal.Conv) error {
+	// Tracks Spanner names that have been used for foreign key constraints
+	// and indexes. We use this to ensure we generate unique names when
+	// we map from Postgres to Spanner since Spanner requires all foreign
+	// key and index names to be distinct (you can't use the same name
+	// for a foreign key constraint and an index).
+	usedNames := make(map[string]bool)
+	// As Spanner uses same namespace for table names, foreign key constraint
+	// names and index names, we need to pre-populate usedNames with Spanner table
+	// names to handle collision with foreign key names and index names.
+	for _, srcTable := range conv.SrcSchema {
+		spTableName, err := internal.GetSpannerTable(conv, srcTable.Name)
+		if err != nil {
+			conv.Unexpected(fmt.Sprintf("Couldn't map source table %s to Spanner: %s", srcTable.Name, err))
+			continue
+		}
+		usedNames[spTableName] = true
+	}
+	for _, srcTable := range conv.SrcSchema {
+		spTableName, err := internal.GetSpannerTable(conv, srcTable.Name)
+		if err != nil {
+			conv.Unexpected(fmt.Sprintf("Couldn't map source table %s to Spanner: %s", srcTable.Name, err))
+			continue
+		}
+		var spColNames []string
+		spColDef := make(map[string]ddl.ColumnDef)
+		conv.Issues[srcTable.Name] = make(map[string][]internal.SchemaIssue)
+		// Iterate over columns using ColNames order.
+		for _, srcColName := range srcTable.ColNames {
+			srcCol := srcTable.ColDefs[srcColName]
+			colName, err := internal.GetSpannerCol(conv, srcTable.Name, srcCol.Name, false)
+			if err != nil {
+				conv.Unexpected(fmt.Sprintf("Couldn't map source column %s of table %s to Spanner: %s", srcTable.Name, srcCol.Name, err))
+				continue
+			}
+			spColNames = append(spColNames, colName)
+			ty, issues := toSpannerType(conv, srcCol.Type.Name, srcCol.Type.Mods)
+			if len(srcCol.Type.ArrayBounds) > 1 {
+				ty = ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+				issues = append(issues, internal.MultiDimensionalArray)
+			}
+			// TODO: add issues for all elements of srcCol.Ignored.
+			if srcCol.Ignored.ForeignKey {
+				issues = append(issues, internal.ForeignKey)
+			}
+			if srcCol.Ignored.Default {
+				issues = append(issues, internal.DefaultValue)
+			}
+			if len(issues) > 0 {
+				conv.Issues[srcTable.Name][srcCol.Name] = issues
+			}
+			ty.IsArray = false // TODO - check if the data conversion is handled fine
```

Comment:
I don't see how this work.
If the source database has array of int, then ty will be INT64. With this code, we will end up with the Spanner type INT64, and that won't work. We need to generate STRING(MAX).

I think you need something like the code at line 67 (for multi-dimensional arrays) i.e something like

if (len(srcCol.Type.ArrayBounds) == 1) {
  ty = ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
}

Reply:
I ran into this during testing. Looking to correct it. Will try this approach.